bitcoin/src/blockdata/script.rs

```diff
@@ -1184,16 +1184,21 @@ impl ScriptBuf {
     /// for a P2WPKH output. The `scriptCode` is described in [BIP143].
     ///
     /// [BIP143]: <https://github.com/bitcoin/bips/blob/99701f68a88ce33b2d0838eb84e115cef505b4c2/bip-0143.mediawiki>
+    // from bip:
+    // > For P2WPKH witness program, the scriptCode is 0x1976a914{20-byte-pubkey-hash}88ac.
     pub fn p2wpkh_script_code(&self) -> Option<ScriptBuf> {
         if !self.is_v0_p2wpkh() {
             return None
         }
         let script = Builder::new()
-            .push_opcode(OP_DUP)
-            .push_opcode(OP_HASH160)
+            // Why don't we push OP_PUSHBYTES_25 (0x19)
```

Comment:
These are two bugs making canceling out each other and making things work. In all cases except code separator(that we explicitly don't support). `script_code` is just serialized `Script` (spending script pubkey to be precise). When serializing `Script`, we prefix it with its length and that is where the 019 (25) comes from. 

According to the BIP, the script_code should also prefix and the sighash function should accept the `script_code` as input. The correct way to fix this would be to change the `segwit_signature_hash` API. 

One suggestion, feel free to improvise:


```rust

pub struct Bip143ScriptCode(Script);

impl Bip143ScriptCode {
    pub fn new_wpkh(wpkh: Script) -> Option<Self>;
    pub fn new_wsh(wsh: Script) -> Option<Self>;
    // Assume the given script as the correctly constructed script code. Useful for custom cases with code separator, or in cases where the user wants to explicitly trust the given script as correctly computed script code.
    pub fn dangerous_assume_script_code(raw_script: Script) -> Self;
}

// New API that takes script code as BIP143ScriptCode instead of raw &Script.
// Deprecate the previous `segwit_signature_hash` API. I wanted a reason to add the `v0` there :P 
pub fn segwit_v0_signature_hash( .. ,script_code: &Bip143ScriptCode, ...)
```

Reply:
> When serializing Script, we prefix it with its length

Aaaah thanks man, I missed that. For any other readers
```rust
impl Encodable for Script {
    #[inline]
    fn consensus_encode<W: io::Write + ?Sized>(&self, w: &mut W) -> Result<usize, io::Error> {
        crate::consensus::encode::consensus_encode_with_size(&self.0, w)
    }
}
```
and 
```rust
pub(crate) fn consensus_encode_with_size<S: io::Write>(data: &[u8], mut s: S) -> Result<usize, io::Error> {
    let vi_len = VarInt(data.len() as u64).consensus_encode(&mut s)?;
    s.emit_slice(data)?;
    Ok(vi_len + data.len())
}
```

Will hack up something based on your suggestion. Thanks again.

Reply:
Isn't this related to the tagged scripts? Wouldn't it be better to implement script tagging first?

Reply:
What is "script tagging"? I'm unable to find any reference to it on the internet.

Reply:
@tcharding parameterizing the `Script` type by a dummy type indicating its purpose (scriptSig, scriptPubKey, witnessScript, taproot/segwit/legacy, etc)

Reply:
ahhh, nice one, I was wondering why we didn't express the script purpose anywhere.

Reply:
Currently your suggestion is implemented @sanket1729 but I'm going to have a crack at script tagging.